program/src/processor.rs

```diff
@@ -644,3 +666,189 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
+            rewards in 0u64..,
+            token_supply in 0u64..,
+        ) {
+            // Calculate.
+            //
+            // For all possible values of rewards and token_supply, the
+            // calculation should never return an error, hence the
+            // `unwrap` here.
+            //
+            // The scaling to `u128` prevents multiplication from breaking
+            // the `u64::MAX` ceiling, and the `token_supply == 0` check
+            // prevents `checked_div` returning `None` from a zero
+            // denominator.
+            let result = calculate_rewards_per_token(rewards, token_supply).unwrap();
+            // Evaluate.
+            if token_supply == 0 {
+                prop_assert_eq!(result, 0);
+            } else {
+                let expected = (rewards as u128)
+                    .checked_mul(REWARDS_PER_TOKEN_SCALING_FACTOR)
+                    .and_then(|product| product.checked_div(token_supply as u128))
+                    .unwrap();
+                prop_assert_eq!(result, expected);
+            }
+        }
+    }
+
+    // Since the marginal rate (current - last) within this function is describing
+    // rewards _per token_, which is calculated by `calculate_rewards_per_token`,
+    // the maximum permitted true reward difference on the system that can be
+    // tolerated by a maximum token account balance (100% of token supply) depends
+    // on the token supply of the mint.
+    //
+    // Consider the calculations for both functions:
+    //
+    //     rewards_per_token = (reward * 1e9) / mint_supply
+    //     eligible_rewards = (marginal_rewards_per_token * balance) / 1e9
+    //
+    // By rearranging the first formula (for `calculate_rewards_per_token`), we
+    // can calculate the total accumulated system-level reward based on the
+    // given rewards per token and mint supply.
+    //
+    //     accumulated_rewards =
+    //              (accumulated_rewards_per_token * mint_supply) / e9
+    //
+    // By rearranging the second formula (which is for this function
+    // `calculate_eligible_rewards`), and plugging in `u64::MAX * 1e9` for
+    // `eligible_rewards` (function upper bound), we can obtain a formula for
+    // the maximum marginal rate (current - last) supported.
+    //
+    //     max_marginal_rewards_per_token = (u64::MAX * 1e9) / balance
+    //
+    // The largest token account balance possible is 100% of the mint supply,
+    // so we can rewrite the rearranged second formula as follows.
+    //
+    //     max_marginal_rewards_per_token = (u64::MAX * 1e9) / mint_supply
+    //
+    // Now, plugging the formula for `max_marginal_rewards_per_token` into the
+    // `accumulated_rewards_per_token` of the rearranged first formula, we get
+    // the following.
+    //
+    //     max_accumulated_rewards = (
+    //                          ((u64::MAX * 1e9) / mint_supply) * mint_supply
+    //                        ) / 1e9
```

Comment:
Sure, I can clear the terminology up. The idea was that it's the maximum marginal rate that could be calculated - say from a holder with 100% of token supply whose last seen rate is zero and the current rate is massive.

Reply:
> Maybe it would be clearer to say max_undistributed_rewards, since "accumulated" implies the total over the course of the system.

This is what I'm implying though. The rearranged `rewards_per_token` function is designed to go from a marginal reward to the system-level using the mint supply. If the maximum marginal reward is our bound, then we are trying to compute the maximum accumulated rewards where this breaks, should a holder have 100% of supply and never claim since inception.